main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,255 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
+          // Optimizations for Char.
+          case (_, Expression.Char(lit1), Expression.Char(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Float32.
+          case (_, Expression.Float32(lit1), Expression.Float32(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Float32(lit1 + lit2)
+              case BinaryOperator.Minus => Expression.Float32(lit1 - lit2)
+              case BinaryOperator.Times => Expression.Float32(lit1 * lit2)
+              case BinaryOperator.Divide => Expression.Float32(lit1 / lit2)
+              case BinaryOperator.Exponentiate => Expression.Float32(scala.math.pow(lit1, lit2).toFloat)
+              case BinaryOperator.Less => toExp(lit1 < lit2)
+              case BinaryOperator.LessEqual => toExp(lit1 <= lit2)
+              case BinaryOperator.Greater => toExp(lit1 > lit2)
+              case BinaryOperator.GreaterEqual => toExp(lit1 >= lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Float64.
+          case (_, Expression.Float64(lit1), Expression.Float64(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Float64(lit1 + lit2)
+              case BinaryOperator.Minus => Expression.Float64(lit1 - lit2)
+              case BinaryOperator.Times => Expression.Float64(lit1 * lit2)
+              case BinaryOperator.Divide => Expression.Float64(lit1 / lit2)
+              case BinaryOperator.Exponentiate => Expression.Float64(scala.math.pow(lit1, lit2))
+              case BinaryOperator.Less => toExp(lit1 < lit2)
+              case BinaryOperator.LessEqual => toExp(lit1 <= lit2)
+              case BinaryOperator.Greater => toExp(lit1 > lit2)
+              case BinaryOperator.GreaterEqual => toExp(lit1 >= lit2)
+              case _ => Expression.Binary(op, e1, e2, tpe, loc)
+            }
+          // Optimizations for Int8.
+          case (_, Expression.Int8(lit1), Expression.Int8(lit2)) =>
+            op match {
+              case BinaryOperator.Equal => toExp(lit1 == lit2)
+              case BinaryOperator.NotEqual => toExp(lit1 != lit2)
+              case BinaryOperator.Plus => Expression.Int8((lit1 + lit2).toByte)
+              case BinaryOperator.Minus => Expression.Int8((lit1 - lit2).toByte)
+              case BinaryOperator.Times => Expression.Int8((lit1 * lit2).toByte)
+              case BinaryOperator.Divide => Expression.Int8((lit1 / lit2).toByte)
```

Comment:
def f: Int = 42 / 0

Ups... Crashes the compiler, but was suppose to crash the program at run-time, not the compiler.

Two choices: In case of division by zero (a) preserve the code that does x / 0, or (b) emit a warning about possible division by zero.

Here (a) is probably the better choice, since the optimizer is not meant to find errors, and it is totally reasonable that you want to write a program that divides by zero (for testing purposes etc.)